File: 117-populating-next-right-pointers-in-each-node-ii/populating-next-right-pointers-in-each-node-ii.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    Node* connect(Node* root) {
        if (!root) return root;

        queue<Node*> q;
        q.push(root);

        while(!q.empty()){
            int n = q.size();
            vector<Node*> v(n);
            for(int i=0; i<n; i++){
                Node* curr = q.front();
                q.pop();
                v[i] = curr;

                if(curr->left) q.push(curr->left);
                if(curr->right) q.push(curr->right);
            }

            for(int i=0; i<n; i++){
                if (i == n-1){
                    Node *curr = v[i];
                    curr->next = nullptr;
                    continue;
                }
                Node *curr = v[i];
                Node *nex = v[i+1];
                curr->next = nex;
            }
        }

        return root;
    }
};

File: 117-populating-next-right-pointers-in-each-node-ii/populating-next-right-pointers-in-each-node-ii.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Left *Node
 *     Right *Node
 *     Next *Node
 * }
 */

func connect(root *Node) *Node {
	if root == nil {
        return root
    }

    var q []*Node
    q = append(q, root)

    for len(q) != 0 {
        n := len(q)
        var level []*Node
        for i := 0; i < n; i++ {
            node := q[0]
            q = q[1:]

            level = append(level, node)

            if node.Left != nil {
                q = append(q, node.Left)
            }
            if node.Right != nil {
                q = append(q, node.Right)
            }
        }

        for i := 0; i < n-1; i++ {
            node := level[i]
            node.Next = level[i+1]
        }
    }

    return root
}